File: server/src/main/kotlin/com/heerkirov/hedge/server/form/Topic.kt
package com.heerkirov.hedge.server.form

import com.heerkirov.hedge.server.library.form.*
import com.heerkirov.hedge.server.model.meta.Topic
import com.heerkirov.hedge.server.utils.types.Opt
import com.heerkirov.hedge.server.utils.types.OrderItem

data class TopicRes(val id: Int, val name: String, val otherNames: List<String>, val keywords: List<String>,
                    val type: Topic.Type, val favorite: Boolean,
                    val annotations: List<Topic.CachedAnnotation>,
                    val score: Int?, val count: Int)

data class TopicSimpleRes(val id: Int, val name: String, val isExported: Boolean)

data class TopicDetailRes(val id: Int, val name: String, val parent: Parent?,
                          val otherNames: List<String>, val keywords: List<String>, val description: String,
                          val type: Topic.Type, val favorite: Boolean,
                          val annotations: List<Topic.CachedAnnotation>,
                          val links: List<Topic.Link>,
                          val score: Int?, val count: Int, val originScore: Int?) {
    data class Parent(val id: Int, val name: String, val type: Topic.Type)
}

data class TopicFilter(@Limit val limit: Int,
                       @Offset val offset: Int,
                       @Search val search: String?,
                       @Order(options = ["id", "name", "score", "count", "createTime", "updateTime"])
                       val order: List<OrderItem>? = null,
                       val type: Topic.Type? = null,
                       val favorite: Boolean? = null,
                       val parentId: Int? = null,
                       val annotationIds: List<Int>? = null)

data class TopicCreateForm(@NotBlank val name: String,
                           val otherNames: List<String>? = null,
                           val parentId: Int? = null,
                           val type: Topic.Type = Topic.Type.UNKNOWN,
                           val keywords: List<String>? = null,
                           val description: String = "",
                           val links: List<Topic.Link>? = null,
                           val annotations: List<Any>? = null,
                           val favorite: Boolean = false,
                           val score: Int? = null)

data class TopicUpdateForm(@NotBlank val name: Opt<String>,
                           val otherNames: Opt<List<String>?>,
                           val parentId: Opt<Int?>,
                           val type: Opt<Topic.Type>,
                           val keywords: Opt<List<String>?>,
                           val description: Opt<String>,
                           val links: Opt<List<Topic.Link>?>,
                           val annotations: Opt<List<Any>?>,
                           val favorite: Opt<Boolean>,
                           val score: Opt<Int?>)

fun newTopicRes(topic: Topic) = TopicRes(topic.id, topic.name, topic.otherNames, topic.keywords, topic.type, topic.favorite, topic.cachedAnnotations ?: emptyList(), topic.exportedScore, topic.cachedCount)

fun newTopicDetailRes(topic: Topic, parent: Topic?) = TopicDetailRes(topic.id, topic.name,
    parent?.let { TopicDetailRes.Parent(it.id, it.name, it.type) },
    topic.otherNames, topic.keywords, topic.description, topic.type, topic.favorite,
    topic.cachedAnnotations ?: emptyList(),
    topic.links ?: emptyList(),
    topic.score, topic.cachedCount, topic.score)

File: frontend/src/functions/adapter-http/impl/topic.ts
import { HttpInstance, Response } from "../server"
import { IdResponse, LimitAndOffsetFilter, Link, ListResult, mapFromOrderList, OrderList } from "./generic"
import { SimpleAnnotation } from "./annotations"

export function createTopicEndpoint(http: HttpInstance): TopicEndpoint {
    return {
        list: http.createQueryRequest("/api/topics", "GET", {
            parseQuery: mapFromTopicFilter
        }),
        create: http.createDataRequest("/api/topics", "POST"),
        get: http.createPathRequest(id => `/api/topics/${id}`),
        update: http.createPathDataRequest(id => `/api/topics/${id}`, "PATCH"),
        delete: http.createPathRequest(id => `/api/topics/${id}`, "DELETE")
    }
}

function mapFromTopicFilter(data: TopicFilter): any {
    return {
        ...data,
        order: mapFromOrderList(data.order),
        annotationsIds: data.annotationIds?.length ? data.annotationIds.join(",") : undefined
    }
}

/**
 * 主题。
 */
export interface TopicEndpoint {
    /**
     * 查询主题列表。
     */
    list(filter: TopicFilter): Promise<Response<ListResult<Topic>>>
    /**
     * 新建主题。
     * @exception ALREADY_EXISTS ("Topic", "name", name) 主题重名
     * @exception NOT_EXISTS ("annotations"|"parentId", id) 指定的资源不存在
     * @exception NOT_SUITABLE ("annotations", id) 指定的资源不适用。对于annotations，此注解的target要求不能应用于此种类的tag
     * @exception RECURSIVE_PARENT 在父标签检查中发现了闭环
     * @exception ILLEGAL_CONSTRAINT ("type", "parent", parentType) 当前主题的类型和父主题的类型不能兼容
     */
    create(form: TopicCreateForm): Promise<Response<IdResponse>>
    /**
     * 查看主题。
     * @exception NOT_FOUND
     */
    get(id: number): Promise<Response<DetailTopic>>
    /**
     * 更改主题。
     * @exception NOT_FOUND
     * @exception ALREADY_EXISTS ("Topic", "name", name) 主题重名
     * @exception NOT_EXISTS ("annotations"|"parentId", id) 指定的资源不存在
     * @exception NOT_SUITABLE ("annotations", id) 指定的资源不适用。对于annotations，此注解的target要求不能应用于此种类的tag
     * @exception RECURSIVE_PARENT 在父标签检查中发现了闭环
     * @exception ILLEGAL_CONSTRAINT ("type", "parent", parentType) 当前主题的类型和父主题的类型不能兼容
     */
    update(id: number, form: TopicUpdateForm): Promise<Response<null>>
    /**
     * 删除主题。
     * @exception NOT_FOUND
     */
    delete(id: number): Promise<Response<null>>
}

export type TopicType = "UNKNOWN" | "COPYRIGHT" | "WORK" | "CHARACTER"

export interface Topic {
    /**
     * topic id。
     */
    id: number
    /**
     * 主题名称。需要遵守tag name规范。
     */
    name: string
    /**
     * 其他名称。需要遵守tag name规范。
     */
    otherNames: string[]
    /**
     * 关键字。需要遵守tag name规范。
     */
    keywords: string[]
    /**
     * 主题类型。
     */
    type: TopicType
    /**
     * 标记为收藏。
     */
    favorite: boolean
    /**
     * 注解。
     */
    annotations: SimpleAnnotation[]
    /**
     * 手写的评分或关联的项目的平均分。
     */
    score: number | null
    /**
     * 关联的项目数量。
     */
    count: number
}

export interface DetailTopic extends Topic {
    /**
     * 主题的父标签。
     */
    parent: ParentTopic | null
    /**
     * 简介。
     */
    description: string
    /**
     * 相关链接。
     */
    links: Link[]
    /**
     * 手写的原始评分。
     */
     originScore: number | null
}

export interface ParentTopic {
    id: number
    name: string
    type: TopicType
}

export interface SimpleTopic {
    id: number
    name: string
    isExported: boolean
}

export interface TopicCreateForm {
    name: string
    otherNames?: string[] | null
    parentId?: number | null
    type?: TopicType
    description?: string
    keywords?: string[]
    links?: Link[] | null
    annotations?: (string | number)[] | null
    favorite?: boolean
    score?: number | null
}

export interface TopicUpdateForm {
    name?: string
    otherNames?: string[] | null
    parentId?: number | null
    type?: TopicType
    description?: string
    keywords?: string[]
    links?: Link[] | null
    annotations?: (string | number)[] | null
    favorite?: boolean
    score?: number | null
}

export type TopicFilter = TopicQueryFilter & LimitAndOffsetFilter

export interface TopicQueryFilter {
    search?: string
    order?: OrderList<"id" | "name" | "score" | "count" | "createTime" | "updateTime">
    type?: TopicType
    favorite?: boolean
    parentId?: number
    annotationIds?: number[]
}
